File: moli-protocol/src/domains/page/tests/same_document_history/cross_origin_navigation.js
async function crossOriginNavigationProbe(api, crossOrigin, destination, action) {
  const frame = document.createElement('iframe');
  frame.name = 'navigation-target';
  const url = new URL('/cross-origin-navigation.html?', location.href);
  if (crossOrigin) url.hostname = location.hostname === 'localhost' ? '127.0.0.1' : 'localhost';
  const initialURL = url.href;
  frame.src = initialURL;
  const records = [], sourceEvents = [], queue = [];
  let resolveMessage, timer, form;
  const onMessage = event => {
    if (event.source !== frame.contentWindow || !event.data?.kind) return;
    if (event.data.kind === 'navigate') records.push(event.data);
    if (resolveMessage) {
      const resolve = resolveMessage;
      resolveMessage = null;
      resolve(event.data);
    } else queue.push(event.data);
  };
  const nextMessage = async () => {
    if (queue.length) return queue.shift();
    try {
      return await Promise.race([
        new Promise(resolve => { resolveMessage = resolve; }),
        new Promise((_, reject) => {
          timer = setTimeout(() => reject(new Error('missing child message')), 2000);
        })
      ]);
    } finally { clearTimeout(timer); }
  };
  const waitFor = async (...kinds) => {
    for (;;) {
      const message = await nextMessage();
      if (kinds.includes(message.kind)) return message;
    }
  };
  const onNavigate = event => sourceEvents.push(event.navigationType);
  addEventListener('message', onMessage);
  navigation.addEventListener('navigate', onNavigate);
  try {
    document.body.append(frame);
    await waitFor('load');
    frame.contentWindow.postMessage({kind: 'configure', action}, '*');
    const initial = await waitFor('configured');
    if (destination === 'fragment') url.hash = 'fragment';
    else url.pathname = '/cross-origin-navigation-destination.html';
    if (destination === 'cross-origin-document') {
      url.hostname = url.hostname === 'localhost' ? '127.0.0.1' : 'localhost';
    }
    let returnedExisting = null;
    if (api === 'href') frame.contentWindow.location.href = url.href;
    else if (api === 'window-location') frame.contentWindow.location = url.href;
    else if (api === 'replace') frame.contentWindow.location.replace(url.href);
    else if (api === 'open') returnedExisting = open(url.href, frame.name) === frame.contentWindow;
    else {
      form = document.createElement('form');
      form.action = url.href;
      form.target = frame.name;
      if (api === 'post') {
        form.method = 'POST';
        form.innerHTML = '<input name="secret" value="source-only">';
      }
      document.body.append(form);
      if (api === 'requestSubmit') form.requestSubmit();
      else form.submit();
    }
    const terminal = action === 'cancel'
      ? await waitFor('navigate', 'load')
      : await waitFor('success', 'load');
    frame.contentWindow.postMessage({kind: 'report'}, '*');
    const report = await waitFor('report');
    return {terminal: terminal.kind, records, sourceEvents, returnedExisting,
      unchanged: report.href === initialURL,
      reached: report.href === url.href,
      retained: report.token === initial.token,
      entriesDelta: report.entries - initial.entries,
      currentEntryMatches: report.currentEntryMatches};
  } finally {
    clearTimeout(timer);
    form?.remove();
    frame.remove();
    removeEventListener('message', onMessage);
    navigation.removeEventListener('navigate', onNavigate);
  }
}

File: moli-protocol/src/domains/page/tests/same_document_history/cross_origin_navigation.rs
use super::*;

const PROBE: &str = include_str!("cross_origin_navigation.js");
const CHILD: &str = include_str!("cross_origin_navigation.html");

async fn check_cross_origin_navigation(api: &str) {
    for cross_origin in [false, true] {
        for destination in ["fragment", "document", "cross-origin-document"] {
            for action in ["cancel", "proceed", "intercept"] {
                if destination == "cross-origin-document" && action == "intercept" {
                    continue;
                }
                let routes = axum::Router::new()
                    .route(
                        "/cross-origin-navigation.html",
                        axum::routing::any(|| async { axum::response::Html(CHILD) }),
                    )
                    .route(
                        "/cross-origin-navigation-destination.html",
                        axum::routing::any(|| async { axum::response::Html(CHILD) }),
                    );
                let mut page = SameDocumentPage::with_routes(routes).await;
                let result = page.evaluate(&format!(
                    "{PROBE}\ncrossOriginNavigationProbe({api:?}, {cross_origin}, {destination:?}, {action:?})"
                )).await;
                let fragment = destination == "fragment" && api != "post";
                let fires = !cross_origin || fragment;
                let canceled = fires && action == "cancel";
                let retained = canceled || fragment || (fires && action == "intercept");
                let form = matches!(api, "submit" | "requestSubmit" | "post");
                let records = if fires {
                    json!([{
                        "kind": "navigate", "type": if api == "replace" { "replace" } else { "push" },
                        "cancelable": true, "canIntercept": destination != "cross-origin-document",
                        "sameDocument": fragment, "hashChange": fragment,
                        "sourceNull": cross_origin || !form,
                        "formData": if api == "post" { json!([["secret", "source-only"]]) } else { json!(null) },
                        "key": "", "id": "", "index": -1,
                    }])
                } else {
                    json!([])
                };
                assert_eq!(
                    result,
                    json!({
                        "terminal": if canceled { "navigate" } else if retained { "success" } else { "load" },
                        "records": records, "sourceEvents": [],
                        "returnedExisting": if api == "open" { json!(true) } else { json!(null) },
                        "unchanged": canceled, "reached": !canceled, "retained": retained,
                        "entriesDelta": if canceled || api == "replace" || destination == "cross-origin-document" { 0 } else { 1 },
                        "currentEntryMatches": true,
                    }),
                    "{api}/{cross_origin}/{destination}/{action}"
                );
            }
        }
    }
}

#[tokio::test(flavor = "multi_thread")]
async fn cross_origin_navigation_location_href() {
    check_cross_origin_navigation("href").await;
}

#[tokio::test(flavor = "multi_thread")]
async fn cross_origin_navigation_window_location() {
    check_cross_origin_navigation("window-location").await;
}

#[tokio::test(flavor = "multi_thread")]
async fn cross_origin_navigation_location_replace() {
    check_cross_origin_navigation("replace").await;
}

#[tokio::test(flavor = "multi_thread")]
async fn cross_origin_navigation_window_open() {
    check_cross_origin_navigation("open").await;
}

#[tokio::test(flavor = "multi_thread")]
async fn cross_origin_navigation_form_submit() {
    check_cross_origin_navigation("submit").await;
}

#[tokio::test(flavor = "multi_thread")]
async fn cross_origin_navigation_form_request_submit() {
    check_cross_origin_navigation("requestSubmit").await;
}

#[tokio::test(flavor = "multi_thread")]
async fn cross_origin_navigation_form_post_stays_cross_document() {
    check_cross_origin_navigation("post").await;
}
